### DISCUSSION TITLE
multiple query on a single page take too long on loading

### DISCUSSION THREAD

**Discussion by samg666** [Budibase/budibase#3225] [Help]
Hell,

I have a page with a main data provider & repeater, and other 4 related repeaters on the tabbed containers. The main repeater button clicked, then will change the multiple app state value, then other 4 data provider/repeaters will be triggered to run the queries and load the new data. My issue, once I click the main repeater button, it takes a little bit too long on loading. I found that it take about 5s on scripting. Is there anything I can do to improve the speed?

<img width="2564" alt="iShot2021-11-02 21 54 04" src="https://user-images.githubusercontent.com/29649651/139861939-d78a790e-592d-4469-84ac-e7165eb3b6ef.png">

**Comment by mjashanks** [comment]
Hi Sam. Would it be possible to send an export of your app to support@budibase.com?

This would include all the data in your internal tables, so I'd understand if you wouldn't want to.

If not...
- please send a screenshot of your component tree (the left-hand panel in the design section)
- are you using much custom javascript? I.e. code that you've written yourself? Could any of those be responsible? If you are only doing basic data manipulation, then it should be fine.
- Are your repeaters nested? This could result in `Repeater1 * Repeater2 * Components ` to be rendered - which could again be compounded by and custom javascript.


Thanks!

**Reply by samg666** [reply]
@mjashanks Hi Michael, please find the screenshot of my component tree. I used most are handlebar, and about 2 very simple javascript(something like `return A+B*C`). I don't have nested repeater(shown as RPT in my diagram), but with many nested data provider(shown as DP in my diagram). I have 7 data provider, and for easier to bind any of the data provider, I put most data providers nested, and repeaters as childs. I use `{{#forEach DP1}}{{field1}}{{/forEach}}` to bind the data on some of the buttons. I am not sure if it's a good design in my case. 

<img width="597" alt="iShot2021-11-05 09 19 17" src="https://user-images.githubusercontent.com/29649651/140444475-cd69c9b9-8fe3-48f4-acb8-9e952379c3ac.png">

**Reply by samg666** [reply]
@mjashanks Hi Michael, as you can see on my 1st screenshot, I use multiple buttons and containers created something like tabbed containers, once a button clicked, the corresponding containers shown and others will be hidden. My question is, if a repeater and data provider inside a hidden container, will the binded query being run? As I have mentioned last post, I have 7 data providers and 5 repeaters on a single page, but in the most case there are just 3 repeaters and 3-5 queries needs to be run, others actually inside the hidden containers, if something inside the hidden container  no need to run then it may save loading time.

**Reply by mjashanks** [reply]
Hey @ggs331 - apologies for the late reply

I think the issue here is that you are nesting Data Providers inside each other. You should only do that if it's absolutely necessary  - i.e. when one has a dependency on the other.

The way would have structured your data providers means that the API requests will get executed sequentially...

```
DP1 
  > then DP2 
    > then DP3 
      > then DP5 
        > then DP5
```

Each one will wait for the last one to finish, before running.

If possible you should keep them all on the same level. This will cause them all to be executed at the same time:

```
> DP 1
> and DP 2
> and DP 3
> and DP 4
> and DP 5
```

Hopefully, I've explained myself well - let me know if not!

**Reply by samg666** [reply]
@mjashanks  Michael, thanks for your update. I will try to re-arrange my data provider, the reason why I put many data nested is because I will need to reference from them, as budibase can only reference from the parents, if I don't put them in the top level, then I will need to create many more data providers. But it seems create more data provider will be a better solution than put many data provier nested.

BTW, if a data provier inside a hidden container, will the data provider being executed?
